adr/2025-04-01-context-gateway.md

```diff
@@ -0,0 +1,395 @@
+    {
+        return self::COMMAND_KEY;
+    }
+}
+```
+
+Example command handler class:
+
+```php
+<?php declare(strict_types=1);
+
+class ChangeCurrencyCommandHandler extends AbstractContextGatewayCommandHandler
+{
+    public function __construct(
+        private readonly EntityRepository $currencyRepository
+    ) {
+    }
+
+    /**
+     * @param ChangeCurrencyCommand $command
+     */
+    public function handle(AbstractContextGatewayCommand $command, SalesChannelContext $context, array &$parameters): void
+    {
+        $criteria = new Criteria();
+        $criteria->addFilter(new EqualsFilter('isoCode', $command->iso));
+
+        $currencyId = $this->currencyRepository->searchIds($criteria, $context->getContext())->firstId();
+
+        if ($currencyId === null) {
+            return;
+        }
+
+        // $parameters will be used by ContextSwitchRoute to update the context
+        $parameters['currencyId'] = $currencyId;
+    }
+
+    public static function supportedCommands(): array
+    {
+        return [ChangeCurrencyCommand::class];
+    }
+```
+
+#### Command Registry
+
+Command handlers will be registered in the `ContextGatewayCommandRegistry` using a registry pattern.
+
+All implementations of `AbstractContextGatewayCommandHandler` must be tagged with the `shopware.context.gateway.command` service tag.
+This ensures they are automatically discovered and made available to the app system for command execution.
+
+#### Event
+
+A new event, `ContextGatewayCommandsCollectedEvent`, will be introduced.
+This event is dispatched after the `AppContextGateway` collects all command responses from registered app servers, but before the commands are executed.
+
+It allows plugins to inspect, modify, or append commands based on the full context and payload provided to the app servers.
+
+## Consequences
+
+### App PHP SDK
+
+The `app-php-sdk` will be extended to support the new gateway and its data contract.
+
+Enhancements include:
+
+- Context gateway requests can be deserialized into a `ContextGatewayAction` object.
+- Responses can easily be created with a `ContextGatewayResponse` object.
+- Each supported context command will have a dedicated class, allowing strongly-typed manipulation and validation of its payload.
+
+### App Bundle (Symfony)
+
+The App Bundle will be updated to support the new gateway endpoint automatically.
+
+Incoming context gateway requests and outgoing responses will be automatically resolved to the appropriate DTOs (`ContextGatewayAction`, `ContextGatewayResponse`),
+making it seamless for app developers to implement handlers using standard Symfony controllers.
+
+## Security Considerations
+
+The `ContextGateway` introduces new responsibilities and associated risks due to its ability to manipulate the `SalesChannelContext`.
+
+Potential concerns include:
+
+- **Customer impersonation**: App servers have the ability to log in existing customers, even without knowing their passwords.
```

Comment:
To concerns we could also add: https://shopware.stoplight.io/docs/admin-api/829793ea23281-generate-a-customer-impersonation-token

This endpoint probably should be deleted after introducing this functionality

Reply:
What does this Store-API functionality have to do with that Admin-API functionality? Yes, you could login via an app gateway, but you would still need an app server. The customer impersonation token is there to login in via a token, that does not require communication with a separate server.